<center>
<img src="../../img/ods_stickers.jpg">
## Открытый курс по машинному обучению
<center>
Автор материала: Юрий Кашницкий, программист-исследователь Mail.Ru Group <br> 

Материал распространяется на условиях лицензии [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/). Можно использовать в любых целях (редактировать, поправлять и брать за основу), кроме коммерческих, но с обязательным упоминанием автора материала.

# <center>Домашнє завдання № 1 (демо).<br> Аналіз даних доходів населення UCI Adult</center>

**В завданні пропонується за допомогою Pandas дати відповідь на кілька запитань по даним репозиторія UCI [Adult](https://archive.ics.uci.edu/ml/datasets/Adult).**

Унікальні значення ознак:
- age: continuous.
- workclass: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
- fnlwgt: continuous.
- education: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.
- education-num: continuous.
- marital-status: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
- occupation: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
- relationship: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.
- race: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
- sex: Female, Male.
- capital-gain: continuous.
- capital-loss: continuous.
- hours-per-week: continuous.
- native-country: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.   
- salary: >50K,<=50K

In [93]:
import pandas as pd

In [94]:
data = pd.read_csv('adult.data.csv')
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

**1. Скільки чоловіків и жінок (ознака *sex*) представлено в цьому наборі даних?**

In [95]:
data["sex"].value_counts()

sex
Male      21790
Female    10771
Name: count, dtype: int64

**2. Який середній вік (ознака *age*) жінок?**

In [96]:
age = data[data["sex"] == "Female"]['age'].mean()
print(f"The average age of female is: {float(age)}")

The average age of female is: 36.85823043357163


**3. Яка доля громадян Німеччини (ознака *native-country*)?**

In [97]:
germans = data[data['native-country'] == "Germany"].shape[0]
total = data.shape[0]
fate = (germans / total) * 100
print(f"The fate of German citizens: {fate}")

The fate of German citizens: 0.42074874850281013


**4-5. Які средні значення та средньоквадратичні відхилення віку тих, хто отримує більше 50K на рік (ознака *salary*) і тих, хто отримує менше 50K на рік?**

In [98]:
more_then = data[data['salary'] == ">50K"]
less_then = data[data['salary'] == "<=50K"]
# counting mean value
mean_more_then = more_then['age'].mean()
mean_less_then = less_then['age'].mean()
# counting standard deviation
sd_more_then = more_then['age'].std()
sd_less_then = less_then['age'].std()

print(f"Mean value and standard deviations of the ages of those who receive more than 50K per year are: {mean_more_then}; {sd_more_then}")
print(f"Mean value and standard deviations of the ages of those who receive less than 50K per year are: {mean_less_then}; {sd_less_then}")

Mean value and standard deviations of the ages of those who receive more than 50K per year are: 44.24984058155847; 10.519027719851826
Mean value and standard deviations of the ages of those who receive less than 50K per year are: 36.78373786407767; 14.02008849082488


Стандартне відхилення: 10.51
Вік осіб, які заробляють понад 50 тис., ближчий до середнього віку (44,25), що вказує на те, що більшість людей у цій групі мають однаковий вік.

Стандартне відхилення: 14.02
Вік осіб, які заробляють менше 50 тис., більш розкиданий, що вказує на більшу різноманітність віку в цій групі.

**6. Чи правда, що люди, які отримують більше 50k, мають як мінімум вищу освіту? (ознака *education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters* или *Doctorate*)**

In [99]:
higher_education = ['Bachelors', 'Prof-school', 'Assoc-acdm', 'Assoc-voc', 'Masters', 'Doctorate']

has_higher_education = more_then['education'].isin(higher_education).all()
print(f"Is it true that people who earn more than 50k have at least a university degree? {has_higher_education}")

Is it true that people who earn more than 50k have at least a university degree? False


**7. Порахуйте статистику віку для кожної раси (ознака *race*) і кожної статі. Використайте *groupby* и *describe*. Знайдіть таким сином максимальний вік чоловіків раси *Amer-Indian-Eskimo*.**

In [100]:
gr = data.groupby(['race', 'sex'])['age'].describe()[['max']].reset_index().copy()
gr

race     sex   max
0  Amer-Indian-Eskimo  Female  80.0
1  Amer-Indian-Eskimo    Male  82.0
2  Asian-Pac-Islander  Female  75.0
3  Asian-Pac-Islander    Male  90.0
4               Black  Female  90.0
5               Black    Male  90.0
6               Other  Female  74.0
7               Other    Male  77.0
8               White  Female  90.0
9               White    Male  90.0

Відповідь: максимальний вік чоловіків раси *Amer-Indian-Eskimo **82 роки** 

**8. Серед кого більша доля багатіших (>50K): серед одружених чи не одружених чоловіків (ознака *marital-status*)? За одружених вважаємо тих, у кого *marital-status* починається із *Married* (Married-civ-spouse, Married-spouse-absent или Married-AF-spouse), всі інші - не одружені.**

In [101]:
# getting all males with salary >50K (with marital-status and sex column)
salary_male = data[(data['sex'] == 'Male') & (data['salary'] == '>50K')][['salary', 'marital-status', 'sex']].copy()
# adding new column marital-status-group that will split all males into 2 groups: Married or not.
salary_male['marital-status-group'] = salary_male['marital-status'].apply(
    lambda x: 'Married' if x.startswith('Married') else 'Not-Married'
)
# getting count of rich males who is married and not.
married_male = salary_male[salary_male['marital-status-group'] == "Married"].shape
unmarried_male = salary_male[salary_male['marital-status-group'] == "Not-Married"].shape
# 
print(f"Share of richer men (>50K) among married is: {married_male[0]} or unmarried men is: {unmarried_male[0]}")

Share of richer men (>50K) among married is: 5965 or unmarried men is: 697


**9. Яку максимальну кількість годин в тиждень працює людина (ознака *hours-per-week*)? Скільки людей працюють таку кількість годин і який відсоток серед них багатих?**

In [102]:
max_hours_per_week = data['hours-per-week'].max()
print(f"Max hours per week: {max_hours_per_week}")
# getting all people who work max hours per week
count_of_people = data[data['hours-per-week'] == max_hours_per_week].copy()
print(f"Number of people who works max hour per week: {count_of_people.shape[0]}")
# counting rich people percentage
rich_percentage = (count_of_people[count_of_people['salary'] == ">50K"].shape[0] / count_of_people.shape[0]) * 100
print(f"Rich percentage: {rich_percentage}")

Max hours per week: 99
Number of people who works max hour per week: 85
Rich percentage: 29.411764705882355


**10. Порахуйте середню тривалість роботи на тиждень (*hours-per-week*) серед багатих і бідних (*salary*) для кожної країни (*native-country*).**

In [103]:
data['status'] = data['salary'].apply(lambda x: 'Rich' if x.startswith('>') else 'Pure')
data.groupby(['native-country', 'status'])['hours-per-week'].mean().reset_index().copy()

native-country status  hours-per-week
0               ?   Pure       40.164760
1               ?   Rich       45.547945
2        Cambodia   Pure       41.416667
3        Cambodia   Rich       40.000000
4          Canada   Pure       37.914634
..            ...    ...             ...
77  United-States   Rich       45.505369
78        Vietnam   Pure       37.193548
79        Vietnam   Rich       39.200000
80     Yugoslavia   Pure       41.600000
81     Yugoslavia   Rich       49.500000

[82 rows x 3 columns]